scripts/bin/update-packages

```diff
@@ -148,22 +151,130 @@ _fetch_and_cache_tags() {
+				__GITHUB_PACKAGES+=("${pkg_dir}")
+			else
+				__PACKAGES+=("${pkg_dir}")
+			fi
 		done
 	done
 
+	local GITHUB_GRAPHQL_QUERIES=() COUNTER=0
+	declare -A __GITHUB_GRAPHQL_PACKAGES=()
+	# Ignore non-constant sources
+	# shellcheck disable=SC1091
+	for pkg_dir in "${__GITHUB_PACKAGES[@]}"; do
+		local PKG_SRCURL TAG_TYPE OWNER REPO
+		read -r PKG_SRCURL TAG_TYPE < <(
+			set +u
+			source "${pkg_dir}/build.sh"
+			echo "${TERMUX_PKG_SRCURL} ${TERMUX_PKG_UPDATE_TAG_TYPE}"
+		)
+
+		read -r OWNER REPO < <(sed -E 's|^(git\+)?https?://github.com/([^/]+)/([^/]+).*|\2 \3|' <<< "$PKG_SRCURL")
+
+		if [[ -z "${TAG_TYPE}" ]]; then # If not set, then decide on the basis of url.
+			if [[ "${PKG_SRCURL:0:4}" == "git+" ]]; then
+				# Get newest tag.
+				TAG_TYPE="newest-tag"
+			else
+				# Get the latest release tag.
+				TAG_TYPE="latest-release-tag"
+			fi
+		fi
+
+		# This holds the source repositories, and query type for each repo we need to query.
+		# Newlines need to be escaped for the query parser
+		GITHUB_GRAPHQL_QUERIES+=( "_$((COUNTER++)): repository(owner: \\\"${OWNER}\\\", name: \\\"${REPO}\\\") { ..._${TAG_TYPE//-/_} }" )
+
+		unset PKG_SRCURL TAG_TYPE OWNER REPO
+	done
+
+	# Batch size for fetching tags, 100 seems to work consistently.
+	local BATCH BATCH_SIZE=100
+	echo "// Begin test: $(date)" > /tmp/query-12345
+	for (( BATCH = 0; ${#GITHUB_GRAPHQL_QUERIES[@]} >= BATCH_SIZE * BATCH ; BATCH++ )); do
+
+		echo "Starting batch $BATCH at: ${GITHUB_GRAPHQL_QUERIES[$BATCH * $BATCH_SIZE]//\\/}"
+
+		# JSON strings cannot contain tabs or newlines
+		# so shutup shellcheck complaining about escapes in single quotes
+		local QUERY
+
+		# Start the GraphQL query with our two fragments for getting the latest tag from a release, and from refs/tags
+		# These are defined only if needed, so this one is for '_latest_release_tag'
+		grep -q '_latest_release_tag' <<< "${GITHUB_GRAPHQL_QUERIES[@]:$BATCH * $BATCH_SIZE:$BATCH_SIZE}" && {
+			QUERY+="$(printf '%s\n' \
+			'fragment _latest_release_tag on Repository {' \
+			'  latestRelease { tagName }' \
+			'}')"
+		}
+
+		grep -q '_newest_tag' <<< "${GITHUB_GRAPHQL_QUERIES[@]:$BATCH * $BATCH_SIZE:$BATCH_SIZE}" && {
+			QUERY+="$(printf '%s\n' \
+			'fragment _newest_tag on Repository {' \
+			'  refs( refPrefix: \"refs/tags/\" first: 1 orderBy: {field: TAG_COMMIT_DATE, direction: DESC}) {' \
+			'    nodes { name }' \
+			'  }' \
+			'}')"
+		}
+
+		QUERY+='query {'
+
+		# Fill out the query body with the package repos we need to query for updates
+		# Lastly fetch the rate limit utilization
+		echo "// Batch: $BATCH" >> /tmp/query-12345
+		printf -v QUERY '%s\n' \
+				"${QUERY}" \
+				"${GITHUB_GRAPHQL_QUERIES[@]:$BATCH * $BATCH_SIZE:$BATCH_SIZE}" \
+				'ratelimit: rateLimit { cost limit remaining used resetAt }' \
+				'}' \
+
+		printf '%s' "${QUERY}" >> /tmp/query-12345
+		response="$(printf '{ "query": "%s" }' "${QUERY//$'\n'/ }" | curl -fsL \
```

Comment:
You mentioned that github does not accept queries with more than 250 version fetch requests. Did you try to put more than one query to a single curl invocation? In this case we will have only one call to github API even if it will cost us 11 graphql points so it should not trigger any simultaneous request restriction checker.

Reply:
250 was an overestimation on my part.
And yes I did try multiple queries in in single POST, that didn't work.
We might be able to do multiple separate POST requests, but that isn't really something I wanna be messing with right now.

100 per request works.
And 11 GraphQL API requests per auto-update cycle keeps us well under the rate limit.

Reply:
As far as I understand the restrictions also include one request per second and requests being non-simultaneous. So probably we should add `sleep 1` here. Or `sleep 0.5` if you want to save 5 secs of CI time. I am not sure how much time the single `curl` invocation takes.